### PR TITLE
Lims5300 4426

### PR DESCRIPTION
**Pull request by ghada-37**
➜  1lims-automation git:(LIMS5300-4426) ✗ nosetests-3.4 -vs --logging-level=WARNING ui_testing/testcases/basic_tests/test05_testplans.py:TestPlansTestCases.test012_create_testplans_same_name_article_materialtype_1_all --tc-file=config.ini
LIMS-3499 [with same='all'] ...         
2020-01-04 02:28:04.391 | INFO     | ui_testing.testcases.base_test:setUp:35 - Test case : test012_create_testplans_same_name_article_materialtype_1_all
2020-01-04 02:28:21.172 | INFO     | api_testing.apis.base_api:info:58 - Get authorized api session.
2020-01-04 02:28:22.177 | INFO     | api_testing.apis.base_api:info:58 - session ID : eyJhbGciOi .....
2020-01-04 02:28:22.178 | INFO     | ui_testing.pages.login_page:login:10 - Login admin : admin.
2020-01-04 02:28:38.271 | INFO     | ui_testing.pages.testplans_page:get_test_plans_page:11 -  + Get test plans page.
2020-01-04 02:28:43.019 | INFO     | ui_testing.pages.base_pages:sleep_small:39 -  Small sleep.
2020-01-04 02:28:48.022 | INFO     | api_testing.apis.base_api:info:58 - GET : https://automation.1lims.com/api/testPlans
2020-01-04 02:28:48.843 | INFO     | api_testing.apis.base_api:info:58 - Status code: 200
2020-01-04 02:28:48.845 | INFO     | ui_testing.pages.testplan_page:create_new_test_plan:72 -  Create new test plan
2020-01-04 02:28:49.442 | INFO     | ui_testing.pages.base_pages:sleep_small:39 -  Small sleep.
2020-01-04 02:28:55.069 | INFO     | ui_testing.pages.testplan_page:create_new_test_plan:79 -  With Raw Material material type
2020-01-04 02:28:57.614 | INFO     | ui_testing.pages.base_pages:sleep_tiny:35 -  Tiny sleep.
2020-01-04 02:28:59.617 | INFO     | ui_testing.pages.testplan_page:create_new_test_plan:87 -  With 6670f6a318 article
2020-01-04 02:29:04.150 | INFO     | ui_testing.pages.base_pages:save:51 - save the changes
2020-01-04 02:29:14.336 | INFO     | ui_testing.pages.testplan_page:create_new_test_plan:99 -  Test plan name : d3f153277a
2020-01-04 02:29:14.337 | INFO     | ui_testing.testcases.basic_tests.test05_testplans:test012_create_testplans_same_name_article_materialtype:409 - New testplan is created successfully with name: d3f153277a, article name: 6670f6a318 and material type: Raw Material
2020-01-04 02:29:14.337 | INFO     | ui_testing.testcases.basic_tests.test05_testplans:test012_create_testplans_same_name_article_materialtype:430 - Attempting to create another testplan with the same data but All Articles
2020-01-04 02:29:14.338 | INFO     | ui_testing.pages.testplan_page:create_new_test_plan:72 -  Create new test plan
2020-01-04 02:29:14.748 | INFO     | ui_testing.pages.base_pages:sleep_small:39 -  Small sleep.
2020-01-04 02:29:20.268 | INFO     | ui_testing.pages.testplan_page:create_new_test_plan:79 -  With Raw Material material type
2020-01-04 02:29:22.815 | INFO     | ui_testing.pages.base_pages:sleep_tiny:35 -  Tiny sleep.
2020-01-04 02:29:24.818 | INFO     | ui_testing.pages.testplan_page:create_new_test_plan:87 -  With All article
2020-01-04 02:29:27.214 | INFO     | ui_testing.pages.base_pages:save:51 - save the changes
2020-01-04 02:29:37.386 | INFO     | ui_testing.pages.testplan_page:create_new_test_plan:99 -  Test plan name : d3f153277a
2020-01-04 02:29:37.387 | INFO     | ui_testing.testcases.basic_tests.test05_testplans:test012_create_testplans_same_name_article_materialtype:436 - Waiting for the error message to make sure that validation forbids the creation of two testplans having the same name, material type and one with ALL article
2020-01-04 02:29:37.435 | INFO     | ui_testing.testcases.basic_tests.test05_testplans:test012_create_testplans_same_name_article_materialtype:441 - Assert the error message to make sure that validation forbids the creation of two testplans having the same name, material type and one with All article? True
2020-01-04 02:29:37.506 | INFO     | ui_testing.testcases.base_test:tearDown:60 - TearDown.     
ok

----------------------------------------------------------------------
Ran 1 test in 93.115s

OK

➜  1lims-automation git:(LIMS5300-4426) ✗ nosetests-3.4 -vs --logging-level=WARNING ui_testing/testcases/basic_tests/test05_testplans.py:TestPlansTestCases.test015_update_quantification_limit_new_version_created --tc-file=config.ini
New: Test plan: Limits of quantification Approach: In case I update the limits ...      
2020-01-04 02:16:15.671 | INFO     | ui_testing.testcases.base_test:setUp:35 - Test case : test015_update_quantification_limit_new_version_created
2020-01-04 02:16:32.481 | INFO     | api_testing.apis.base_api:info:58 - Get authorized api session.
2020-01-04 02:16:33.468 | INFO     | api_testing.apis.base_api:info:58 - session ID : eyJhbGciOi .....
2020-01-04 02:16:33.470 | INFO     | ui_testing.pages.login_page:login:10 - Login admin : admin.
2020-01-04 02:16:51.604 | INFO     | ui_testing.pages.testplans_page:get_test_plans_page:11 -  + Get test plans page.
2020-01-04 02:16:56.565 | INFO     | ui_testing.pages.base_pages:sleep_small:39 -  Small sleep.
2020-01-04 02:17:01.571 | INFO     | api_testing.apis.base_api:info:58 - GET : https://automation.1lims.com/api/testPlans
2020-01-04 02:17:02.410 | INFO     | api_testing.apis.base_api:info:58 - Status code: 200
2020-01-04 02:17:02.412 | INFO     | ui_testing.pages.testunits_page:get_test_units_page:11 -  + Get test units page.
2020-01-04 02:17:05.987 | INFO     | ui_testing.pages.base_pages:sleep_small:39 -  Small sleep.
2020-01-04 02:17:10.991 | INFO     | ui_testing.pages.testunit_page:create_quantitative_testunit:72 - create_quantitative_testunit
2020-01-04 02:17:10.992 | INFO     | ui_testing.pages.testunit_page:click_create_new_testunit:15 - Click Create New Test Unit
2020-01-04 02:17:11.434 | INFO     | ui_testing.pages.base_pages:sleep_tiny:35 -  Tiny sleep.
2020-01-04 02:17:13.436 | INFO     | ui_testing.pages.testunit_page:set_testunit_name:128 - Set testunit name to be: 65b474d9d3
2020-01-04 02:17:13.705 | INFO     | ui_testing.pages.testunit_page:set_material_type:100 - Set material type to be "Raw Material", if it is empty, then it will be random
2020-01-04 02:17:16.228 | INFO     | ui_testing.pages.testunit_page:set_category:114 - Set category to be "", if it is empty, then it will be random
2020-01-04 02:17:18.652 | INFO     | ui_testing.pages.testunit_page:get_category:124 - Get Testunit category
2020-01-04 02:17:18.734 | INFO     | ui_testing.pages.testunit_page:set_testunit_type:158 - Set testunit type to be Quantitative
2020-01-04 02:17:21.232 | INFO     | ui_testing.pages.testunit_page:use_specification_or_quantification:167 - Check to use quan
2020-01-04 02:17:21.656 | INFO     | ui_testing.pages.base_pages:sleep_tiny:35 -  Tiny sleep.
2020-01-04 02:17:23.659 | INFO     | ui_testing.pages.testunit_page:set_quan_upper_limit:204 - Set quantification upper limit to be 53692
2020-01-04 02:17:23.916 | INFO     | ui_testing.pages.testunit_page:set_quan_lower_limit:208 - Set quantification lower limit to be 53692
2020-01-04 02:17:24.157 | INFO     | ui_testing.pages.testunit_page:set_quan_unit:212 - Set quantification unit to be 
2020-01-04 02:17:24.502 | INFO     | ui_testing.pages.testunit_page:set_testunit_iteration:144 - Set testunit iterations to be: 1
2020-01-04 02:17:24.725 | INFO     | ui_testing.pages.testunit_page:set_method:7 - Set testunit method to be: e2cbe08b6f
2020-01-04 02:17:24.985 | INFO     | ui_testing.pages.base_pages:save:51 - Save new testunit
2020-01-04 02:17:35.140 | INFO     | ui_testing.pages.testplans_page:get_test_plans_page:11 -  + Get test plans page.
2020-01-04 02:17:38.559 | INFO     | ui_testing.pages.base_pages:sleep_small:39 -  Small sleep.
2020-01-04 02:17:43.565 | INFO     | ui_testing.pages.testplans_page:get_test_plan_edit_page:24 - Navigating to testplan bc288a590e edit page
2020-01-04 02:17:54.927 | INFO     | ui_testing.pages.base_pages:sleep_small:39 -  Small sleep.
2020-01-04 02:18:00.417 | INFO     | ui_testing.pages.base_pages:sleep_small:39 -  Small sleep.
2020-01-04 02:18:10.220 | INFO     | ui_testing.pages.base_pages:sleep_tiny:35 -  Tiny sleep.
2020-01-04 02:18:12.222 | INFO     | ui_testing.pages.base_pages:save:51 - save the changes
2020-01-04 02:18:22.384 | INFO     | ui_testing.pages.testplans_page:get_test_plans_page:11 -  + Get test plans page.
2020-01-04 02:18:26.165 | INFO     | ui_testing.pages.base_pages:sleep_small:39 -  Small sleep.
2020-01-04 02:18:31.170 | INFO     | ui_testing.pages.testplans_page:get_test_plan_edit_page:24 - Navigating to testplan bc288a590e edit page
2020-01-04 02:18:42.493 | INFO     | ui_testing.pages.base_pages:sleep_small:39 -  Small sleep.
2020-01-04 02:18:47.494 | INFO     | ui_testing.pages.testplan_page:navigate_to_testunits_selection_page:107 - Navigating to testplan create/update step 2
2020-01-04 02:18:47.685 | INFO     | ui_testing.pages.base_pages:sleep_small:39 -  Small sleep.
2020-01-04 02:18:53.080 | INFO     | ui_testing.pages.base_pages:sleep_small:39 -  Small sleep.
2020-01-04 02:18:58.086 | INFO     | ui_testing.pages.base_pages:save:51 - save the changes
2020-01-04 02:19:08.298 | INFO     | ui_testing.pages.testplan_page:save_and_confirm_popup:145 - Accepting the changes made
2020-01-04 02:19:08.510 | INFO     | ui_testing.pages.base_pages:sleep_small:39 -  Small sleep.
2020-01-04 02:19:13.516 | INFO     | ui_testing.pages.testplans_page:get_test_plans_page:11 -  + Get test plans page.
2020-01-04 02:19:16.692 | INFO     | ui_testing.pages.base_pages:sleep_small:39 -  Small sleep.
2020-01-04 02:19:21.698 | INFO     | ui_testing.testcases.basic_tests.test05_testplans:test015_update_quantification_limit_new_version_created:514 - Getting the currently changed testplan to check its status and version
2020-01-04 02:19:35.279 | INFO     | ui_testing.pages.base_pages:sleep_small:39 -  Small sleep.
2020-01-04 02:19:40.522 | INFO     | ui_testing.pages.base_pages:sleep_medium:43 -  Medium sleep.
2020-01-04 02:19:51.290 | INFO     | ui_testing.testcases.basic_tests.test05_testplans:test015_update_quantification_limit_new_version_created:520 - Asserting that the Quantification limits updated correctly
2020-01-04 02:19:51.290 | INFO     | ui_testing.testcases.basic_tests.test05_testplans:test015_update_quantification_limit_new_version_created:522 - the Quantification limits updated correctly
2020-01-04 02:19:51.365 | INFO     | ui_testing.testcases.base_test:tearDown:60 - TearDown.     
ok

----------------------------------------------------------------------
Ran 1 test in 215.696s

OK

### CODE REVIEW

**Comment by 0xIslamTaha** [comment]
@ghada-37 travis and conflict

File: ui_testing/pages/testplan_page.py [hunk 1]
```diff
@@ -240,3 +240,13 @@ def get_testunit_quantification_limit(self, testunits, testunit_display_name):
                 quantification_limit = testunit['Quantification Limit']
                 return quantification_limit
 
+    def edit_test_unit_quantification_limits(self, upper, lower):
+        self.navigate_to_testunits_selection_page()
+        elems = self.base_selenium.find_elements('general:col_6')
+        old_upper = self.base_selenium.find_element_in_element(source=elems[4], destination_element='general:input')
+        old_lower = self.base_selenium.find_element_in_element(source=elems[5], destination_element='general:input')
```
Comment (0xIslamTaha) [inline ui_testing/pages/testplan_page.py]: Move elements values to the elements.py file

File: ui_testing/pages/testplan_page.py [hunk 2]
```diff
@@ -240,3 +240,13 @@ def get_testunit_quantification_limit(self, testunits, testunit_display_name):
                 quantification_limit = testunit['Quantification Limit']
                 return quantification_limit
 
+    def edit_test_unit_quantification_limits(self, upper, lower):
```
Comment (0xIslamTaha) [inline ui_testing/pages/testplan_page.py]: I'm not sure, but maybe there is already implemented functions to do this

File: ui_testing/pages/testplan_page.py [hunk 3]
```diff
@@ -240,3 +240,13 @@ def get_testunit_quantification_limit(self, testunits, testunit_display_name):
...
+        self.navigate_to_testunits_selection_page()
+        elems = self.base_selenium.find_elements('general:col_6')
+        old_upper = self.base_selenium.find_element_in_element(source=elems[4], destination_element='general:input')
+        old_lower = self.base_selenium.find_element_in_element(source=elems[5], destination_element='general:input')
+        old_lower.clear()
+        old_upper.clear()
+        old_upper.send_keys(upper)
+        old_lower.send_keys(lower)
```
Comment (0xIslamTaha) [inline ui_testing/pages/testplan_page.py]: user `https://github.com/modeso-open-source/1lims-automation/blob/master/ui_testing/pages/base_selenium.py#L310`

File: ui_testing/testcases/basic_tests/test05_testplans.py [hunk 4]
```diff
@@ -385,37 +386,60 @@ def test011_test_plan_completed_to_inprogress(self):
...
+                                                article='All')
+
+            self.base_selenium.LOGGER.info(
+                'Waiting for the error message to make sure that validation forbids the creation of two testplans having the same name, material type and one with ALL article')
+            validation_result = self.base_selenium.wait_element(element='general:oh_snap_msg')
+
+            self.base_selenium.LOGGER.info(
+                'Assert the error message to make sure that validation forbids the creation of two testplans having the same name, material type and one with All article? {}'.format(
+                    validation_result))
+            self.assertTrue(validation_result)
```
Comment (0xIslamTaha) [inline ui_testing/testcases/basic_tests/test05_testplans.py]: I can see that the only difference between "if" and "else" code is the value of "article" in the `self.test_plan.create_new_test_plan` so, just do this 
`article = testplan['article'])[0] if same == "same" else "All"` then pass this `article` to the `create_new_test_plan` method

File: ui_testing/testcases/basic_tests/test05_testplans.py [hunk 5]
```diff
@@ -446,41 +470,57 @@ def test013_create_testplans_same_name_different_materialtype(self):
...
+        # create quantitative test unit with quantification limit
+        self.test_unit_page.get_test_units_page()
+        new_name = self.generate_random_string()
+        new_method = self.generate_random_string()
+        new_random_limit = self.generate_random_number()
+        self.test_unit_page.create_quantitative_testunit(name=new_name, material_type=material_type, category='',
+                                                         upper_limit=new_random_limit, lower_limit=new_random_limit,
+                                                         spec_or_quan='quan', method=new_method)
+
+        self.test_unit_page.save(save_btn='general:save_form', logger_msg='Save new testunit')
```
Comment (0xIslamTaha) [inline ui_testing/testcases/basic_tests/test05_testplans.py]: Use API to create it

File: ui_testing/testcases/basic_tests/test05_testplans.py [hunk 6]
```diff
@@ -446,41 +470,57 @@ def test013_create_testplans_same_name_different_materialtype(self):
...
+        # navigate to the chosen test plan edit page
+        self.test_plan.get_test_plans_page()
+        self.test_plan.get_test_plan_edit_page(testplan_name)
+        # navigate to the test units selection tab
+        self.test_plan.set_test_unit(test_unit=new_name)
+        self.test_plan.sleep_tiny()
```
Comment (0xIslamTaha) [inline ui_testing/testcases/basic_tests/test05_testplans.py]: Use API to add a test unit to an existing test plan

File: ui_testing/pages/testplan_page.py (hunk 1 above)
Comment (ghada-37) [reply ui_testing/pages/testplan_page.py]: elements already in elements.py file

File: ui_testing/pages/testplan_page.py (hunk 2 above)
Comment (ghada-37) [reply ui_testing/pages/testplan_page.py]: https://github.com/modeso-open-source/1lims-automation/blob/LIMS5300-4426/ui_testing/pages/testplan_page.py#L232
??

File: ui_testing/testcases/basic_tests/test05_testplans.py (hunk 4 above)
Comment (ghada-37) [reply ui_testing/testcases/basic_tests/test05_testplans.py]: done

File: ui_testing/pages/testplan_page.py (hunk 2 above)
Comment (ghada-37) [reply ui_testing/pages/testplan_page.py]: I used it

File: ui_testing/pages/testplan_page.py (hunk 3 above)
Comment (ghada-37) [reply ui_testing/pages/testplan_page.py]: I removed it and used https://github.com/modeso-open-source/1lims-automation/blob/LIMS5300-4426/ui_testing/pages/testplan_page.py#L232

File: ui_testing/testcases/basic_tests/test05_testplans.py (hunk 6 above)
Comment (ghada-37) [reply ui_testing/testcases/basic_tests/test05_testplans.py]: I couldn't use it

File: ui_testing/testcases/basic_tests/test05_testplans.py (hunk 5 above)
Comment (ghada-37) [reply ui_testing/testcases/basic_tests/test05_testplans.py]: I don't undersatnd create_quantitative_testunit in api file, I need to understand this function, who can help me ?

File: ui_testing/pages/testplan_page.py (hunk 1 above)
Comment (0xIslamTaha) [reply ui_testing/pages/testplan_page.py]: my bad

File: ui_testing/pages/testplan_page.py (hunk 2 above)
Comment (0xIslamTaha) [reply ui_testing/pages/testplan_page.py]: no no dont use the DB one

File: ui_testing/pages/testplan_page.py (hunk 3 above)
Comment (0xIslamTaha) [reply ui_testing/pages/testplan_page.py]: dont do that

File: ui_testing/pages/testplan_page.py (hunk 2 above)
Comment (ghada-37) [reply ui_testing/pages/testplan_page.py]: sorry I mean this function 
https://github.com/modeso-open-source/1lims-automation/blob/LIMS5300-4426/ui_testing/pages/testplan_page.py#L236
def update_upper_lower_limits_of_testunit(self, old_upper, old_lower):
I used it and it worked well, thank you for your note

File: ui_testing/pages/testplan_page.py (hunk 3 above)
Comment (ghada-37) [reply ui_testing/pages/testplan_page.py]: sorry again , I used https://github.com/modeso-open-source/1lims-automation/blob/LIMS5300-4426/ui_testing/pages/testplan_page.py#L236
def update_upper_lower_limits_of_testunit(self, old_upper, old_lower):
I used it and it worked well, thank you for your note

**Pull request closed without merging**